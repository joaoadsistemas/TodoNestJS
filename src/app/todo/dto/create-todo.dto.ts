import { ApiProperty } from "@nestjs/swagger";
import { IsNotEmpty } from "class-validator";

export class CreateTodoDto {

    @IsNotEmpty()
    //shows in swagger that this property should be used
    @ApiProperty()
    task: string;

    @IsNotEmpty()
    //shows in swagger that this property should be used
    @ApiProperty()
    isDone: number;
}