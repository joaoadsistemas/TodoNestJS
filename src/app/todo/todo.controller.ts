
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';
import { TodoService } from './todo.service';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';
import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { IndexTodoSwagger } from './swagger/index-todo.swagger';
import { CreateTodoSwagger } from './swagger/create-todo.swagger';
import { ShowTodoSwagger } from './swagger/show-todo.swagger';
import { UpdateTodoSwagger } from './swagger/update-todo.swagger';    
import { BadRequestSwagger } from './helpers/bad-request.swagger';
import { NotFoundSwagger } from './helpers/not-found.swagger';

@Controller('api/v1/todos')
// change the title name in swagger
@ApiTags('todos')
export class TodoController {
  constructor(private readonly todoService: TodoService) {}

  @Get(':id')
  // Used to describe in swagger what it does
  @ApiOperation({ summary: 'Exibir os dados de uma tarefa' })
  @ApiResponse({ status: 200, description: "Dados de uma tarefa retornada com sucesso", type: ShowTodoSwagger})
  @ApiResponse({ status: 404, description: "Tarefa não foi encontrada", type: NotFoundSwagger })
  async findById(@Param('id', new ParseUUIDPipe()) id: string) {
    return await this.todoService.findOneById(id);
  }




  @Get()
  @ApiOperation({ summary: 'Listar todas as tarefas' })
  @ApiResponse({ status: 200, description: "Dados das tarefas retornada com sucesso", type: IndexTodoSwagger, isArray: true})
  async findAll() {
    return await this.todoService.findAll();
  }



  @Post()
  @ApiOperation({ summary: 'Criar uma nova tarefa' })
  @ApiResponse({ status: 201, description: "Nova tarefa criada com sucesso", type: CreateTodoSwagger })
  @ApiResponse({ status: 400, description: "Parâmetros inválidos", type: BadRequestSwagger })
  async create(@Body() body: CreateTodoDto) {
    return await this.todoService.create(body);
  }



  @Put(':id')
  @ApiOperation({ summary: 'Atualizar os dados de uma tarefa'})
  @ApiResponse({ status: 201, description: "Tarefa atualizada com sucesso", type: UpdateTodoSwagger })
  @ApiResponse({ status: 400, description: 'Dados inválidos', type: BadRequestSwagger })
  @ApiResponse({ status: 404, description: "Tarefa não foi encontrada", type: NotFoundSwagger })
  async update(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() body: UpdateTodoDto,
  ) {
    return await this.todoService.update(id, body);
  }



  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Deletar uma tarefa' })
  @ApiResponse({ status: 204, description: "Tarefa removida com sucesso" })
  @ApiResponse({ status: 404, description: "Tarefa não foi encontrada", type: NotFoundSwagger })
  async delete(@Param('id', new ParseUUIDPipe()) id: string) {
    await this.todoService.delete(id);
  }
}
