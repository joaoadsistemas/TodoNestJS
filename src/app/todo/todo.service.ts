import { Injectable, NotFoundException } from '@nestjs/common';
import { Repository } from 'typeorm';
import { TodoEntity } from './entities/todo.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';

@Injectable()
export class TodoService {

    constructor(
        @InjectRepository(TodoEntity) 
        private readonly todoRepository: Repository<TodoEntity>) {

        } ;


        async findAll() {
            return await this.todoRepository.find();
        }

        async findOneById(id: string) {
            
            try {
                return await this.todoRepository.findOneByOrFail({id});
            } catch (error) {
                throw new NotFoundException(error.message);
            }
            };

        async create(data: CreateTodoDto) {
            return await this.todoRepository.save(this.todoRepository.create(data));
        }

        async update(id: string, data: UpdateTodoDto) {
            const todo = await this.findOneById(id);

            this.todoRepository.merge(todo, data);

            return await this.todoRepository.save(todo);
        }

        async delete(id: string) {
            const todo = await this.findOneById(id);
            this.todoRepository.softDelete(todo);
        }


}
