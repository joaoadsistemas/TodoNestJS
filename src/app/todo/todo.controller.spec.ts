import { Repository } from 'typeorm';
import { Test, TestingModule } from '@nestjs/testing';
import { TodoController } from './todo.controller';
import { TodoService } from './todo.service';
import { TodoEntity } from './entities/todo.entity';
import exp from 'constants';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';



const todoEntityList: Array<TodoEntity> = [
  new TodoEntity({
    id: '1',
    task: 'task-1',
    isDone: 0
  }),
  new TodoEntity({
    id: '2',
    task: 'task-2',
    isDone: 0
  }),
  new TodoEntity({
    id: '3',
    task: 'task-3',
    isDone: 0
  }),
  new TodoEntity({
    id: '4',
    task: 'task-4',
    isDone: 0
  })
]

const newTodoEntity = new TodoEntity({task: 'new-task', isDone: 0});

const updateTodoEntity = new TodoEntity({ task: 'task-1',
isDone: 1 }); 

describe('TodoController', () => {
  let todoController: TodoController;
  let todoService: TodoService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [TodoController],
      providers: [
        {
          provide: TodoService,
          useValue: {
            findAll: jest.fn().mockResolvedValue(todoEntityList),
            create: jest.fn().mockResolvedValue(newTodoEntity),
            update: jest.fn().mockResolvedValue(updateTodoEntity),
            delete: jest.fn().mockResolvedValue(undefined),
            findOneById: jest.fn().mockResolvedValue(todoEntityList[0])
          }
        }
      ]
    }).compile();

    todoController = module.get<TodoController>(TodoController);
    todoService = module.get<TodoService>(TodoService);
  });

  test('should be defined', () => {
    expect(todoController).toBeDefined();
    expect(todoService).toBeDefined();
  });

  describe('findAll', () => {
    it ('should return a todo list entity successfully', async () => {
      // Act
        const result = await todoController.findAll();
      // Assert
      expect(result).toEqual(todoEntityList);
      expect(typeof result).toEqual('object');
    })
  })


  describe('create', () => {
    it('should create a new todo item successfully', async () => {
      // Arrange
      const body: CreateTodoDto = {
        task: 'new task',
        isDone: 0
      }


      // Act
      const result = await todoController.create(body)

      // Assert
      expect(result).toEqual(newTodoEntity);
      expect(todoService.create).toHaveBeenCalledTimes(1);
      expect(todoService.create).toHaveBeenCalledWith(body);
    });


  })

  describe('findById', () => {
    it('should get a todo item successfully', async () => {
      // Act
      const result = await todoController.findById('1');

      //Assert
      expect(result).toEqual(todoEntityList[0])
      expect(todoService.findOneById).toHaveBeenCalledTimes(1);
      expect(todoService.findOneById).toHaveBeenCalledWith('1');
    });


  });


  describe('update', () => {
    it('should update a todo item successfully', async () => {
      // Arrange
      const body: UpdateTodoDto = {
        task: 'task-1',
        isDone: 1
      }

      // Act
      const result = await todoController.update('1', body)

      // Assert
      expect(result).toEqual(updateTodoEntity);
      expect(todoService.update).toHaveBeenCalledTimes(1);
      expect(todoService.update).toHaveBeenCalledWith('1', body);
    })


  });

  describe('delete', () => {
    it('should delete a todo item successfully', async () => {
      // Act
      const result = await todoController.delete('1');
  
      // Assert
      expect(result).toBeUndefined();
    });
  })})
