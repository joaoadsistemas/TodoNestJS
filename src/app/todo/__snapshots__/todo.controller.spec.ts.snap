// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoController delete should throw an exception 1`] = `""`;

exports[`TodoController findAll should throw an exception 1`] = `""`;

exports[`TodoController should throw an exception 1`] = `""`;

exports[`TodoController update should throw an exception 1`] = `""`;
